experimentation/src/main/java/com/automattic/android/experimentation/remote/ExperimentRestClient.kt

```diff
@@ -0,0 +1,47 @@
+package com.automattic.android.experimentation.remote
+
+import com.automattic.android.experimentation.domain.Assignments
+import com.automattic.android.experimentation.domain.Clock
+import com.automattic.android.experimentation.domain.SystemClock
+import com.automattic.android.experimentation.remote.AssignmentsDtoMapper.toAssignments
+import com.squareup.moshi.Moshi
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import java.io.IOException
+
+internal class ExperimentRestClient(
+    private val okHttpClient: OkHttpClient = OkHttpClient(),
+    private val moshi: Moshi = Moshi.Builder().build(),
+    private val jsonAdapter: AssignmentsDtoJsonAdapter = AssignmentsDtoJsonAdapter(moshi),
+    private val urlBuilder: UrlBuilder = ExPlatUrlBuilder(),
+    private val clock: Clock = SystemClock(),
+) {
+
+    suspend fun fetchAssignments(
+        platform: String,
+        experimentNames: List<String>,
+        anonymousId: String? = null,
+    ): Result<Assignments> {
+        val url = urlBuilder.buildUrl(platform, experimentNames, anonymousId)
+
+        val request = Request.Builder()
+            .url(url)
+            .get()
+            .build()
+
+        return withContext(Dispatchers.IO) {
+            okHttpClient.newCall(request).execute().use { response ->
+                if (!response.isSuccessful) {
+                    Result.failure<Assignments>(IOException("Unexpected code $response"))
+                } else {
+                    runCatching {
+                        val dto = jsonAdapter.fromJson(response.body!!.source())!!
```

Comment:
I think `runCatching` is a good fit because it returns the `Result` type - just what we need here. It protects against the NPE (from `!!`) but also from an incorrect JSON we parse here.

> Is there a way for us to get rid of !! with runCatching or you don't mind as anything will flow through the default failure when that happens anyway? 🤔

I don't mind much here about it 😅 I think it's okay to wrap here all exceptions - I understand it's usually an anti-pattern, but we don't want ever to have uncaught exception in `fetchAssignments`

Reply:
Let's the move forward with this solution! 👍 

> ...but also from an incorrect JSON we parse here.

💯